### ISSUE TITLE
Real-time object detection on Raspberry Pi 3

### ISSUE THREAD

**Issue by Eddie1293** [ultralytics/ultralytics#7647]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello @glenn-jocher , these days I've trained an object detection model that I'd like to use in real-time on a Raspberry Pi 3 Model B. I followed your instructions to set up YOLOv8 on the Raspberry Pi, and everything works great. 

However, when I start real-time detection, after about 20 minutes of streaming, the model stops working. At the beginning, I receive a warning message suggesting the use of the `stream=True` parameter for long detections. However, when I include this parameter, the stream doesn't start at all, and I don't receive any error messages. The code I'm using is as follows:

`from ultralytics import YOLO`
`model=YOLO('best.pt')`
`modelmodel(source=0, show=True, conf=0.8, stream=True)`

I am using a USB webcam directly connected to the Raspberry Pi, and I have read in previous discussions that for this type of webcam, the `stream=True` parameter is not needed. However, if I don't use it, the stream stops, and if I include it in the previous code, the stream doesn't start at all, and I don't receive any error messages. Plus, I have already performed all the checks related to the webcam, and it is recognized correctly.

What can I do? Thank you for help!

**Comment by glenn-jocher** [comment]
@Eddie1293 hello! It's great to hear that you've successfully trained your object detection model and are implementing it on a Raspberry Pi 3 Model B. 🎉

The issue you're experiencing with the `stream=True` parameter could be related to the hardware limitations of the Raspberry Pi 3, as real-time detection is resource-intensive. When you don't use `stream=True`, the system might be running out of resources over time, leading to the stream stopping. On the other hand, when you do use `stream=True`, there might be an underlying issue that isn't producing an error message.

For troubleshooting, please ensure that your Raspberry Pi's software is up to date and that you're using the latest version of YOLOv8. Also, monitor the system resources (CPU, memory) during the stream to check for potential bottlenecks.

If the problem persists, consider opening an issue on the Ultralytics YOLOv8 repo with detailed information about your setup and any logs that could help diagnose the problem. Our community and team are always here to help. For further guidance on real-time detection, you can refer to our documentation at https://docs.ultralytics.com.

Keep us posted on your progress, and good luck with your real-time detection project!

**Comment by Eddie1293** [comment]
Hi @glenn-jocher, I've a question regarding the interpretation of the graphs I obtain at the end of the training phase. 
![results](https://github.com/ultralytics/ultralytics/assets/156955243/e66d4ef6-401f-4378-87eb-265844e2b2f1)
In the image I've attached, I'm having trouble understanding what the graphs of metrics/precision(B) and metrics/recall(B) refer to. Are these two graphs related to the training or validation phase? The same question applies to the two graphs related to mAP. Can you help me? Thank you so much!

**Comment by glenn-jocher** [comment]
@Eddie1293 hello! The graphs you're referring to display the performance metrics over the course of training. The metrics/precision(B) and metrics/recall(B) graphs typically represent the precision and recall on the validation dataset, not the training data. The same applies to the mAP (mean Average Precision) graphs; they show the mAP calculated on the validation set. These metrics help you understand how well your model is likely to perform on unseen data. For a more detailed explanation of these metrics, please refer to our documentation at https://docs.ultralytics.com. Keep up the good work! 🚀

**Comment by Eddie1293** [comment]
Thank you very much, @glenn-jocher!!

**Comment by glenn-jocher** [comment]
You're welcome, @Eddie1293! If you have any more questions, feel free to ask. Happy to help! 😊👍

**Comment by manar-majdi** [comment]
Hello @glenn-jocher when i work with this code :
from ultralytics import YOLO
model=YOLO('best.pt')
modelmodel(source=0, show=True, conf=0.8, stream=True)
it give this error : failed to read umages from 0

**Comment by glenn-jocher** [comment]
Hello! It looks like there might be a typo in your code that's causing the issue. The method should be `model.predict()` rather than `modelmodel()`. Here’s the corrected version:

```python
from ultralytics import YOLO
model = YOLO('best.pt')
model.predict(source=0, show=True, conf=0.8, stream=True)
```

Make sure your webcam is properly connected and accessible as `source=0`. If the problem persists, verifying the webcam's connection and availability on your system might help. Let us know how it goes! 😊

**Comment by leesan2000** [comment]
Hi @glenn-jocher I am as well working with a raspberry Pi 3b+, running a YoloV11 nano custom trained network. What would be the next steps to appropiately optimize this network in order for the model detection to run smoother? should I convert it to NCNN? If so, how can I perform this? Thanks

**Comment by glenn-jocher** [comment]
You can optimize your YOLO11 nano model for smoother inference on the Raspberry Pi 3B+ by converting it to NCNN format, as NCNN is highly optimized for mobile and embedded platforms. To export your model, use the `model.export(format="ncnn")` command in Python. You can find detailed instructions in the documentation here: https://docs.ultralytics.com/integrations/ncnn/.
